namer/namer.cc

```diff
@@ -15,16 +18,1336 @@
 
 using namespace std;
 
+// TODO:
+// - Remove redundancy whenever possible.
+// - ENFORCE that NameFinder never produces errors.
+
 namespace sorbet::namer {
 
+namespace {
+class FoundNames;
+
+struct FoundClassRef;
+struct FoundClass;
+struct FoundStaticField;
+struct FoundTypeMember;
+struct FoundMethod;
+
+enum class NameKind : u1 {
+    None = 0,
+    Root = 1,
+    Class = 2,
+    ClassRef = 3,
+    Method = 4,
+    StaticField = 5,
+    TypeMember = 6,
+    Symbol = 7,
+};
```

Comment:
These are not really name kinds, as the elements of this enum look like they describe symbol kinds. Also, we already have a structure called NameKind:

https://github.com/sorbet/sorbet/blob/53b31c158b159c92a2d569295828a6224010b2c9/core/Names.h#L14-L18

Can we give this enum a different name?

Reply:
Maybe `DefinitionKind`?

Reply:
Good point. `DefinitionKind` might be better. It's not quite right for `ClassRef`, which is a reference to a class that may or may not already exist, but it can end up defining classes anyway.